examples.md

```diff
@@ -0,0 +1,178 @@
+	msging "app/gen"
+)
+
+func init() {
+	a := MsgingImpl{}
+	msging.SetExportsWasiMessagingMessagingGuest(a)
+}
+
+type MsgingImpl struct {
+}
+
+func (e MessagingImpl) Configure msging.Result[msging.WasiMessagingConsumerGuestConfiguration, msging.WasiMessagingMessagingGuestError] {
+	// Configure the messaging system
+}
+
+func (e MessagingImpl) Handler msging.Result[struct{}, msging.WasiMessagingMessagingGuestError] {
+	// Handle the message
+}
+
+func main() {}
+```
+
+## Example 2: Guest Usage
+
+### Rust
+
+```rust
+use crate::messaging_types::{GuestConfigurationResult, MessageResult, connect, disconnect};
+
+wit_bindgen::generate!({
+    path: "../wit",
+});
+
+struct MyGuest;
+
+impl guest::Guest for MyGuest {
+    fn configure() -> Result<GuestConfigurationResult, u32> {
+        // This function will be called by the host, who will be maintaining a 
+        // long-lived client connection to a broker or other messaging system.
+        // The client will be subscribed to channels a, b, and c) with no extra configuration.
+        // As soon as configuration is set, the host should kill the Wasm instance.
+        Ok(GuestConfigurationResult {
+            channels: vec!["a".to_string(), "b".to_string(), "c".to_string()],
+            ..Default::default()
+        })
+    }
+
+    fn handler(ms: Vec<MessageResult>) -> Result<(), u32> {
+        // Whenever a message is received on a subscribed channel (from configure()),
+        // the host will call this function. Once the message has been handled,
+        // the host is expected to kill the Wasm instance.
+        
+        for m in ms {
+
+            // match on message metadata for channel name
+            match m.metadata {
+                Some(metadata) => {
+                    for (k, v) in metadata {
+                        if k == "channel" {
+                            match v.as_str() {
+                                "a" => {
+                                    // handle message from channel a
+                                    // [...]
+
+                                    // unsubscribe from channel a
+                                    update_guest_configuration(GuestConfigurationResult {
+                                        channels: vec!["b".to_string(), "c".to_string()],
+                                        ..Default::default()
+                                    }).unwrap()
+
+                                    // abandon message
+                                    consumer::abandon_message(m).unwrap();
+                                }
+                                "b" => {
+                                    // handle message from channel b
+                                    // [...]
+
+                                    // request-reply from channel d
+                                    let client = connect("some-broker").unwrap();
+                                    let msgs = subscribe_try_receive(client, "d", 100).unwrap();
```

Comment:
You have a "subscribe" gesture here but you have a "change configuration" gesture a bit further up that also seems to be modifying what you subscribe to, so I'm confused.

there are several concepts that are quite universal in messaging across brokers and APIs that you need clear mappings for:

a) Config/Options/Credentials -- a set of information required to locate and connect with the broker and supply upfront options such as prefetch buffers and the like
b) Connection - a representation of the network connection including, if provided by the abstracted API or implementation, its connection state
c) Session - a representation of a session or link layered over the network connection. Protocols that have the ability to reconnect (MQTT and AMQP for instance) have such abstractions that allow maintaining a relationship between the parties in spite of the connection dropping out underneath. You also have that for HTTP as a key concept.
d) Link/Subscription/Sender/Receiver - a representation of a particular communication route to an entity, usually either for sending or receiving, sometimes coupled with a volatile subscription that is established with the link to that entity. Sometimes a "subscription" you attach to is an already existing entity in the broker (durable subscription, consumer group/partition). Mind that a client may not be authorized to create such a subscription.

Reply:
> You have a "subscribe" gesture here but you have a "change configuration" gesture a bit further up that also seems to be modifying what you subscribe to, so I'm confused.

`subscribe_try_receive` sets an ephemeral subscription to receive from a channel for a specified amount of time. `update_guest_configuration` can add new subscriptions or update anything that was set in the guest's `configure`.

About the rest the rest of your comment, like mentioned in the `wit/consumer.wit`, `update_guest_configuration` is a:
```
    /// 'Fit-all' type function for updating a guest's configuration – this could be useful for:
    ///     - unsubscribing from a channel,
    ///     - checkpointing,
    ///     - etc..
```

I hadn't given sessions much thought, but I did intend it to cover points (a), (b), and (d). This could be temporary, and a first step to breaking it apart onto other functions down the road.